---
title: eQTL summary statistics
date: "`r Sys.Date()`"
---

```{r global_setting, include = FALSE}
library(data.table)
library(tidyverse)
library(patchwork)
source("Util-rmd.R")

fig.dir = 'Fig/eQTL/'
dir.create(fig.dir, recursive = TRUE, showWarnings = FALSE)

knitr::opts_knit$set(eval.after = 'fig.cap')
knitr::opts_chunk$set(warning = FALSE, message = FALSE, fig.path = fig.dir)
knitr::opts_chunk$set(fig.width=8, fig.height=8, results = "asis", echo = FALSE)
options(stringsAsFactors = FALSE)
FIG.CAP = '**Fig.**'
```

```{sh download_gwas_catalog}
if ! [ -f gwas_catalog_v1.0-associations.tsv.gz ] ; then
	wget https://www.ebi.ac.uk/gwas/api/search/downloads/full
	cat full | gzip -c > gwas_catalog_v1.0-associations.tsv.gz
fi
```

```{r read_gwas_genes_info}
gwas.catalog <-
    read.gwas.catalog("gwas_catalog_v1.0-associations.tsv.gz") %>%
    na.omit

gene.info <- fread("../result/step1/gene.info.gz", header=TRUE, sep = "\t")
ct.order.dt <- read.celltype.order()
```

```{r pseudobulk_statistics}
.file <- ".result_eqtl_pb_stat.rdata"
if(!file.exists(.file)) {
    read.pb.stat <- function(ct) {
        .hdr <- str_c("../result/step5/pseudobulk/", ct)
        .cols <- fread(str_c(.hdr, ".mu_cols.gz"), header=FALSE) %>% unlist
        ret <- fread(str_c(.hdr, ".sum.gz"), header=FALSE, col.names=.cols)
        .hdr <- str_c("../result/step5/filtered/", ct)
        .rows <- fread(str_c(.hdr, ".rows.gz"), header=FALSE, col.names="V1")
        .rows[, c("ensembl_gene_id", "hgnc_symbol") := tstrsplit(V1, "_")]
        .rows[, V1 := NULL]
        ret <- data.table(mean = apply(ret, 1, mean),
                          tot = apply(ret, 1, sum),
                          sd = apply(ret, 1, sd))
        ret <- cbind(.rows, ret) %>% mutate(celltype=ct)
        return(ret)
    }
    ct.pb.stat <-
        lapply(ct.order.dt$celltype, read.pb.stat) %>%
        do.call(what = "rbind")
    save(ct.pb.stat, file = .file)
}
load(.file) # --> ct.pb.stat
```


## Reading and curating eQTL results

```{r}
pip.cutoff <- .5
lodds.cutoff <- log(pip.cutoff/(1-pip.cutoff))
```

```{r read_strong_eqtl_result}
lenient.cutoff <- log(0.05/0.95)

.file <- ".result_eqtl_stat.rdata"

read.eqtl.stat <- function(.file, .cutoff) {
    require(data.table)
    require(tidyverse)
    .cmd <- str_c("gzip -cd ",.file," | awk '$NF > ", .cutoff,"'")
    fread(cmd = .cmd)
}

.eqtl.files <- str_c("../result/step7/chr", 1:22, "_stat.bed.gz")

if(!file.exists(.file)){
    cl <- parallel::makeCluster(10)
    .eqtl.stat <- parallel::parLapply(cl, .eqtl.files,
                                      read.eqtl.stat,
                                      .cutoff = lenient.cutoff) %>%
        do.call(what = rbind)
    parallel::stopCluster(cl)
    save(.eqtl.stat, file = .file)
}

load(.file) # -> .eqtl.stat

.eqtl.stat <- left_join(.eqtl.stat, gene.info) %>%
    mutate(pip = sigmoid(lodds)) %>%
    mutate(lb = abs(`stop` - transcript_start.hg19)) %>%
    mutate(ub = abs(`stop` - transcript_end.hg19)) %>%
    mutate(dist = pmin(lb, ub)) %>%
    as.data.table

.pve.stat <- fread("../result/step7/gene_pve.bed.gz", sep = "\t")
```

## Results

### Multivariate eQTL calling results

```{r fig.width=3, fig.height=6}
.eqtl.count <- .eqtl.stat[, .(n = .N),
                          by = .(floor(pip*20)/20,
                                 floor(-log10(pmax(p.val, 1e-10))))] %>%
    rename(pip = `floor`) %>%
    rename(p.val = `floor.1`)

.pip.scale <-
    scale_x_continuous("multivariate posterior probability",
                       labels = num.round,
                       breaks = seq(.05, .95, .1))

.nsnp.scale <- scale_y_continuous("# SNPs",
                                  labels = function(x) round(x*x),
                                  breaks = sqrt(c(1e3,1e4,5e4,1e5,5e5,1e6)))

p1 <-
    .gg.plot(.eqtl.count[, .(n = sum(`n`)), by = .(pip)]) +
    .pip.scale +
    .nsnp.scale +
    geom_segment(aes(x = pip, xend = pip, y = 0, yend = sqrt(`n`)), size=1.5, colour = "gray") +
    theme(axis.text.x = element_text(angle=90, vjust=1, hjust=1))

.nsnp.fill <-
    scale_fill_distiller("# SNPs",
                         labels = function(x) round(10^x),
                         direction = 1,
                         palette = "PuRd",
                         breaks = log10(c(1, 10, 1e3, 1e5)))

.pval.scale <- scale_y_continuous("univariate p-value",
                                  labels = function(x) num.sci(10^(-x)),
                                  breaks = seq(0,10,2))

p2 <-
    .gg.plot(.eqtl.count, aes(x = pip, y = p.val, fill = log10(n))) +
    geom_tile() +
    theme(axis.text.x = element_text(angle=90, vjust=1, hjust=1)) +
    .nsnp.fill +
    .pval.scale +
    .pip.scale

.dist.count <-
    .eqtl.stat[, .(n = .N),
               by = .(floor(pip*20)/20,
                      floor(dist/1e5)*1e2)] %>%
    rename(pip = `floor`) %>%
    rename(dist = `floor.1`)

.kb.scale <-
    scale_y_continuous("distance (kb)", labels = num.int)

p3 <-
    .gg.plot(.dist.count, aes(x = pip, y = dist, fill = log10(n))) +
    geom_tile() +
    theme(axis.text.x = element_text(angle=90, vjust=1, hjust=1)) +
    .kb.scale +
    .nsnp.fill +
    .pip.scale

plt <- p1 / p2 / p3
print(plt)
.file <- fig.dir %&% "/Fig_eqtl_stat.pdf"
.gg.save(file = .file, plot = plt, width = 3, height = 6)
```

### eQTLs identified by multivariate effect size is different from those identified by univariate analysis

```{r fig.width=3, fig.height=8}
.gene.stat <-
    .eqtl.stat[,
               .(pip = max(pip), p.val = min(p.val)),
               by = .(hgnc_symbol, ensembl_gene_id, celltype)] %>%
    left_join(.pve.stat)

.dt.pip <- .gene.stat[, .(n = .N), by = .(floor(pip*20)/20)] %>%
    rename(pip = `floor`)

.x.scale <- scale_x_continuous("multivariate posterior probability",
                               labels = num.round)

p1 <-
    .gg.plot(.dt.pip, aes(x = pip, y = n)) +
    .x.scale +
    scale_y_continuous("# genes", labels = num.int) +
    geom_bar(stat="identity", size = 1) +
    geom_vline(xintercept = pip.cutoff, size = .5, colour = "red", lty = 2)


p2 <-
    .gg.plot(.gene.stat, aes(x = pip, y = -log10(pmax(p.val, 1e-10)))) +
    geom_hex(aes(colour = ..count..)) +
    geom_vline(xintercept = pip.cutoff, size = .5, colour = "white", lty = 2) +
    .x.scale +
    .pval.scale +
    scale_colour_distiller(trans = "log10", guide="none", direction = 1, palette = "PuRd") +
    scale_fill_distiller(trans = "log10", direction = 1, palette = "PuRd")

p3 <-
    .gg.plot(.gene.stat, aes(x = pip, y = pve)) +
    geom_hex(aes(colour = ..count..)) +
    geom_vline(xintercept = pip.cutoff, size = .5, colour = "white", lty = 2) +
    ylab("PVE by cis-variants") +
    .x.scale +
    scale_colour_distiller(trans = "log10", guide="none", direction = 1, palette = "PuRd") +
    scale_fill_distiller(trans = "log10", direction = 1, palette = "PuRd")

.dt.combined <- .gene.stat %>% left_join(ct.pb.stat)
.dt.combined[is.na(`tot`), `tot` := 0]
.dt.combined[is.na(`mean`), `mean` := 0]
.dt.combined[is.na(`sd`), `sd` := 0]

p4 <-
    .gg.plot(.dt.combined, aes(x = pip, log2(1 + `mean`))) +
    geom_hex(aes(colour = ..count..)) +
    geom_vline(xintercept = pip.cutoff, size = .5, colour = "white", lty = 2) +
    scale_y_continuous("average expression", labels = function(x) num.int(pmax(2^(x) - 1, 0)), breaks = log2(1+ c(1, 100, 1e3, 1e4, 1e5))) +
    .x.scale +
    scale_colour_distiller(trans = "log10", guide="none", direction = 1, palette = "PuRd") +
    scale_fill_distiller(trans = "log10", direction = 1, palette = "PuRd")

plt <- p1 / p2 / p3 / p4
print(plt)
.file <- fig.dir %&% "/Fig_gene_stat.pdf"
.gg.save(file = .file, plot = plt, width = 3, height = 8)
```

### How many variants per gene (with PIP cutoff > `r pip.cutoff`)?

We identified `r num.int(nrow(.eqtl.stat[lodds>lodds.cutoff]))` significant eQTL variants on `r num.int(.n(.eqtl.stat[lodds>lodds.cutoff]$ensembl_gene_id))` genes across `r length(unique(.eqtl.stat[lodds>lodds.cutoff]$celltype))` cell types.

```{r}
.ct.sum <- .eqtl.stat[lodds > lodds.cutoff,
                      .(nv = .n(`#chr` %&% ":" %&% `stop`), ngene = .n(hgnc_symbol)),
                      by = .(celltype)]

.gene.sum <- .eqtl.stat[lodds > lodds.cutoff,
                        .(nv = .n(`#chr` %&% ":" %&% `stop`),
                          nct = .n(celltype)),
                        by = .(hgnc_symbol)]

.var.sum <- .eqtl.stat[lodds > lodds.cutoff,
                       .(ng = .n(hgnc_symbol),
                         nct = .n(celltype),
                         dist = pmin(abs(`stop` - `transcript_start.hg19`),
                                     abs(`stop` - `transcript_end.hg19`))),
                       by = .(`#chr`, `stop`)]
```


```{r fig.width=3, fig.height=4}
.df <- .ct.sum %>%
    mutate(ct = factor(celltype, ct.order.dt$celltype, ct.order.dt$ct.name))

p1 <-
    .gg.plot(.df, aes(x = ct, y = nv, fill = ct)) +
    ylab("number of variants") +
    geom_bar(stat="identity") +
    scale_fill_manual(values = ct.order.dt$hex) +
    theme(legend.position = "none") +
    theme(axis.text.x = element_blank()) +
    theme(axis.title.x = element_blank())

p2 <-
    .gg.plot(.df, aes(x = ct, y = ngene, fill = ct)) +
    ylab("number of genes") +
    xlab("cell types") +
    geom_bar(stat="identity") +
    scale_fill_manual(values = ct.order.dt$hex) +
    theme(legend.position = "none") +
    theme(axis.text.x = element_text(angle=90, vjust=.5, hjust=1))

plt <- p1/p2
print(plt)
.file <- fig.dir %&% "/Fig_celltype_count_eQTL_eGene.pdf"
.gg.save(file = .file, plot = plt, width = 3, height = 4)
```

### Multiple causal variants correspond to multiple cell types

```{r fig.width=3, fig.height=3}
.df <- .gene.sum[, .(nct.m = mean(nct), nct.s = sd(nct)), by = .(nv)]

.aes <- aes(x=nv, y=nct.m, ymin = nct.m - nct.s, ymax = nct.m + nct.s)

plt <-
    .gg.plot(na.omit(.df), .aes) +
    xlab("# associated genetic variants (PIP > " %&% pip.cutoff %&% ")") +
    ylab("# unique cell types") +
    geom_abline(slope=1, color="gray", lty=2, size = .5) +
    geom_linerange(size=.5) +
    geom_point(pch=21, fill="white", stroke=.5)

print(plt)
.file <- fig.dir %&% "/Fig_gene_summary.pdf"
.gg.save(file = .file, plot = plt, width = 3, height = 3)
```

```{r fig.width=3, fig.height=3}
.df <- .var.sum[, .(nct.m = mean(nct), nct.s = sd(nct)), by = .(ng)]

.aes <- aes(x=ng, y=nct.m, ymin = nct.m - nct.s, ymax = nct.m + nct.s)

plt <-
    .gg.plot(na.omit(.df), .aes) +
    xlab("# target genes (PIP > " %&% pip.cutoff %&% ")") +
    ylab("# unique cell types") +
    geom_abline(slope=1, color="gray", lty=2, size = .5) +
    geom_linerange(size=.5) +
    geom_point(pch=21, fill="white", stroke=.5)

print(plt)
.file <- fig.dir %&% "/Fig_var_summary.pdf"
.gg.save(file = .file, plot = plt, width = 3, height = 3)
```

## Overlapping with GWAS genes


### Show cell-type-specific enrichment of GWAS genes

```{r run_gwas_gene_test}
.file <- ".result_eqtl_gwas.rdata"
if(!file.exists(.file)) {
    gwas.gene.test.dt <-
        lapply(ct.order.dt$celltype, function(ct){
            .eqtl.stat[lodds > lodds.cutoff & celltype == ct] %>%
                dplyr::select(ensembl_gene_id) %>%
                unique %>%
                unlist %>%
                run.goseq.gwas(.gwas = gwas.catalog) %>%
                mutate(celltype = ct) %>%
                as.data.table
        }) %>%
        do.call(what = rbind) %>%
        dplyr::select(-under_represented_pvalue) %>%
        dplyr::rename(trait = category, p.val = over_represented_pvalue) %>%
        as.data.table

    save(gwas.gene.test.dt, file = .file)
}
load(.file) ## -> gwas.gene.test.dt
gwas.gene.test.dt[, q.val := p.adjust(p.val, "fdr"), by = .(celltype)]
```

* `r num.int(.n(gwas.gene.test.dt$trait))` traits

```{r fig.width=8, fig.height=3.5}
.traits <-
    gwas.gene.test.dt %>%
    dplyr::filter(q.val < .5) %>%
    group_by(trait) %>%
    slice(which.min(p.val)) %>%
    ungroup() %>%
    arrange(p.val) %>%
    head(70) %>%
    dplyr::select(trait) %>%
    unique %>%
    unlist

.gwas.show <-
    gwas.gene.test.dt[trait %in% .traits] %>%
    mutate(row = celltype, col = trait, weight = -log10(q.val)) %>%
    col.order(.ro = rev(ct.order.dt$celltype), ret.tab = TRUE)

.gwas.sig <- .gwas.show[q.val < .2]

.scale.size <- scale_size_continuous("q-value",
                                     range=c(.1,3),
                                     breaks=-log10(c(1, 0.5, 0.1, 0.05, 0.01)),
                                     labels= function(x) num.sci(10^(-x)))

.scale.fill <- scale_fill_distiller("# overlapped genes",
                                    palette = "PuRd",
                                    direction = 1,
                                    trans="log10",
                                    breaks = c(1, 10, 100, 500))

.aes <- aes(y = row, x = col,
            size = pmin(weight, 2),
            fill = numDEInCat)

plt <-
    .gg.plot(.gwas.show, .aes) +
    ylab("cell type") +
    scale_x_discrete("GWAS traits", position = "top") +
    .scale.fill +
    .scale.size +
    theme(axis.text.x = element_text(angle=90, vjust=0, hjust=0)) +
    geom_point(pch=22, stroke = .2) +
    geom_text(aes(label=numDEInCat), data = .gwas.sig, size=1)

print(plt)
.file <- fig.dir %&% "/Fig_gwas_eGene.pdf"
.gg.save(file = .file, plot = plt, width = 8, height = 3.5)
```

### Assigning cell type specificity to GWAS genes


```{r fig.width = 8, fig.height = 3}
show.gwas.genes <- function(.trait) {

    .df <-
        gwas.catalog[trait == .trait & chr %in% 1:22] %>%
        left_join(.gene.stat) %>%
        na.omit %>%
        mutate(row = celltype, col = hgnc_symbol, weight = logit(pip)) %>%
        col.order(.ro = rev(ct.order.dt$celltype), ret.tab = TRUE)

    .df.freq <- .gene.stat[pip > pip.cutoff,
                           .(n = .n(hgnc_symbol)),
                           by = .(celltype)] %>% 
        mutate(row = factor(celltype, rev(ct.order.dt$celltype)))

    .df.freq.found <-
        .df[pip > pip.cutoff,
            .(n = .n(hgnc_symbol)),
            by = .(celltype)] %>% 
        mutate(row = factor(celltype, rev(ct.order.dt$celltype)))

    p1 <-
        .gg.plot(.df, aes(x = col, y = row, fill = pip)) +
        geom_tile() +
        theme(axis.text.x = element_blank()) +
        theme(axis.ticks.x = element_blank()) +
        theme(legend.position = "left") +
        scale_fill_distiller("PIP", direction = 1) +
        ylab("cell types") +
        xlab(.n(.df$hgnc_symbol) %&% " " %&% .trait %&% " GWAS genes")

    p2 <-
        ggplot(.df.freq, aes(y = row, yend = row, x = 0, xend = `n`)) +
        theme_void() +
        scale_y_discrete(position = "right") +
        scale_x_discrete(position = "top") +
        geom_segment(size = 2, colour = "gray") +
        geom_text(aes(label = num.int(`n`), x = `n`), size = 2, hjust = 1)

    p2.1 <-
        ggplot(.df.freq.found, aes(y = row, yend = row, x = 0, xend = `n`)) +
        theme_void() +
        scale_y_discrete(position = "right") +
        scale_x_discrete(position = "top") +
        geom_segment(size = 2, colour = "gray") +
        geom_text(aes(label = num.int(`n`), x = `n`), size = 2, hjust = 1)

    plt <- (p1 | p2 | p2.1) + plot_layout(nrow = 1, widths = c(5, 1, .5))
    return(plt)
}

for(.trait in .traits){
    plt <- show.gwas.genes(.trait)
    print(plt)
    .code <- str_replace_all(.trait, "[ ']+", "_")
    .file <- fig.dir %&% "/Fig_gwas_eGene_example_" %&% .code %&% ".pdf"
    print(.file)
    .gg.save(file = .file, plot = plt, width = 8, height = 3)
    cat("\n\n")
}
```



## Overlapping gene ontology categories


```{r run_eqtl_go_enrichment}
.file <- ".result_eqtl_go.rdata"

if(!file.exists(.file)) {
    eqtl.go.test.dt <-
        lapply(ct.order.dt$celltype, function(ct){

            .ensg <- .eqtl.stat[, .(ensembl_gene_id)]

            .de.ensg <- .eqtl.stat[lodds > lodds.cutoff & celltype == ct,
                                   .(ensembl_gene_id)]

            run.goseq(.de.ensg, .ensg) %>%
                dplyr::mutate(celltype = ct) %>%
                as.data.table
        }) %>%
    do.call(what = rbind) %>%
    dplyr::select(-under_represented_pvalue) %>%
    dplyr::rename(trait = category, p.val = over_represented_pvalue) %>%
    as.data.table
    save(eqtl.go.test.dt, file = .file)
}

load(.file) ## -> eqtl.go.test.dt
eqtl.go.test.dt[, q.val := p.adjust(p.val, "fdr"), by = .(celltype)]
```

### Biological process

```{r fig.width=10, fig.height=4}

.dt <- eqtl.go.test.dt[`ontology` == "BP",] %>%
    mutate(term.1 = str_remove(term, "^positive regulation of ")) %>%
    mutate(term.1 = str_remove(term.1, "^negative regulation of ")) %>%
    mutate(term.1 = str_remove(term.1, "^regulation of ")) %>%
    mutate(col = str_sub(term.1, 1, 55))

.terms <-
    .dt %>%
    dplyr::filter(numDEInCat >= 10) %>%
    dplyr::filter(numInCat <= 100) %>%
    dplyr::filter(numDEInCat/numInCat > .1) %>%
    group_by(celltype) %>%
    top_n(5, -log10(p.val)) %>%
    ungroup()

.terms.sig <-
    .dt %>%
    dplyr::filter(numDEInCat >= 10) %>%
    dplyr::filter(numInCat <= 100) %>%
    dplyr::filter(numDEInCat/numInCat > .1) %>%
    dplyr::filter(q.val < .1)

.df <-
    .dt[col %in% c(.terms$col, .terms.sig$col)] %>%
    mutate(row = celltype) %>%
    group_by(row, col) %>%
    slice(which.min(p.val)) %>%
    ungroup() %>%
    mutate(weight = -log10(p.val)) %>%
    col.order(.ro = rev(ct.order.dt$celltype), ret.tab=TRUE) %>%
    as.data.table

.df.sig <- .df[q.val < .1]

.scale.size <- scale_size_continuous("p-value",
                                     range=c(0,3.5),
                                     breaks=0:4,
                                     labels=function(x)num.sci(10^(-x)))

.scale.fill <- scale_fill_distiller("%captured",
                                    breaks=seq(0,90,30),
                                    palette="PuRd",
                                    direction = 1)

.aes <- aes(y = row, x = col, size = weight, fill = 100*numDEInCat/numInCat)

plt <-
    .gg.plot(.df, .aes) +
    ylab("cell type") +
    scale_x_discrete("GO term", position = "top") +
    theme(axis.text.x = element_text(angle=90, vjust=0, hjust=0)) +
    geom_point(pch = 22, stroke = .2) +
    geom_text(aes(label=numDEInCat), data = .df.sig, size=1.5) +
    .scale.size +
    .scale.fill

print(plt)
.file <- fig.dir %&% "/Fig_GO_BP_eGene.pdf"
.gg.save(file = .file, plot = plt, width = 10, height = 4)
```


```{r fig.width=11.5, fig.height=4}

.dt <- eqtl.go.test.dt[`ontology` == "BP",] %>%
    mutate(term.1 = str_remove(term, "^positive regulation of ")) %>%
    mutate(term.1 = str_remove(term.1, "^negative regulation of ")) %>%
    mutate(term.1 = str_remove(term.1, "^regulation of ")) %>%
    mutate(col = str_sub(term.1, 1, 55))

.terms <-
    .dt %>%
    dplyr::filter(numDEInCat >= 10) %>%
    dplyr::filter(numInCat >= 100, numInCat <= 1000) %>%
    dplyr::filter(numDEInCat/numInCat > .1) %>%
    group_by(celltype) %>%
    top_n(5, -log10(p.val)) %>%
    ungroup()

.terms.sig <-
    .dt %>%
    dplyr::filter(numDEInCat >= 10) %>%
    dplyr::filter(numInCat >= 100, numInCat <= 1000) %>%
    dplyr::filter(numDEInCat/numInCat > .1) %>%
    dplyr::filter(q.val < .1)

.df <-
    .dt[col %in% c(.terms$col, .terms.sig$col)] %>%
    mutate(row = celltype) %>%
    group_by(row, col) %>%
    slice(which.min(p.val)) %>%
    ungroup() %>%
    mutate(weight = -log10(p.val)) %>%
    col.order(.ro = rev(ct.order.dt$celltype), ret.tab=TRUE) %>%
    as.data.table

.df.sig <- .df[q.val < .1]

.scale.size <- scale_size_continuous("p-value",
                                     range=c(0,3),
                                     breaks=0:4,
                                     labels=function(x)num.sci(10^(-x)))

.scale.fill <- scale_fill_distiller("%captured",
                                    breaks=seq(0,90,30),
                                    palette="PuRd",
                                    direction = 1)

.aes <- aes(y = row, x = col, size = weight, fill = 100*numDEInCat/numInCat)

plt <-
    .gg.plot(.df, .aes) +
    ylab("cell type") +
    scale_x_discrete("GO term", position = "top") +
    theme(axis.text.x = element_text(angle=90, vjust=0, hjust=0)) +
    geom_point(pch = 22, stroke = .2) +
    geom_text(label="*", data = .df.sig, size=1.5) +
    .scale.size +
    .scale.fill

print(plt)
.file <- fig.dir %&% "/Fig_GO_eGene_BP_large.pdf"
.gg.save(file = .file, plot = plt, width = 11.5, height = 4)
```


### Cellular component

```{r fig.width=10, fig.height=4}

.dt <- eqtl.go.test.dt[`ontology` == "CC",] %>%
    mutate(col = str_sub(term, 1, 55))

.terms <-
    .dt %>%
    dplyr::filter(numDEInCat >= 10) %>%
    dplyr::filter(numInCat <= 500) %>%
    dplyr::filter(numDEInCat/numInCat > .1) %>%
    group_by(celltype) %>%
    top_n(5, -log10(p.val)) %>%
    ungroup()

.terms.sig <-
    .dt %>%
    dplyr::filter(numDEInCat >= 10) %>%
    dplyr::filter(numInCat <= 500) %>%
    dplyr::filter(numDEInCat/numInCat > .1) %>%
    dplyr::filter(q.val < .1)

.df <-
    .dt[col %in% c(.terms$col, .terms.sig$col)] %>%
    mutate(row = celltype) %>%
    group_by(row, col) %>%
    slice(which.min(p.val)) %>%
    ungroup() %>%
    mutate(weight = -log10(p.val)) %>%
    col.order(.ro = rev(ct.order.dt$celltype), ret.tab=TRUE) %>%
    as.data.table

.df.sig <- .df[q.val < .1]

.scale.size <- scale_size_continuous("p-value",
                                     range=c(0,3.5),
                                     breaks=0:4,
                                     labels=function(x)num.sci(10^(-x)))

.scale.fill <- scale_fill_distiller("%captured",
                                    breaks=seq(0,90,30),
                                    palette="PuRd",
                                    direction = 1)

.aes <- aes(y = row, x = col, size = weight, fill = 100*numDEInCat/numInCat)

plt <-
    .gg.plot(.df, .aes) +
    ylab("cell type") +
    scale_x_discrete("GO term", position = "top") +
    theme(axis.text.x = element_text(angle=90, vjust=0, hjust=0)) +
    geom_point(pch = 22, stroke = .2) +
    geom_text(aes(label=numDEInCat), data = .df.sig, size=1.5) +
    .scale.size +
    .scale.fill

print(plt)
.file <- fig.dir %&% "/Fig_GO_CC_eGene.pdf"
.gg.save(file = .file, plot = plt, width = 10, height = 4)
```

## Examples

```{r}
.herit.genes <-
    .gene.stat[pip > .5 & p.val < 1e-8, .(hgnc_symbol)] %>%
    unique %>% 
    unlist

.examples <-
    gwas.catalog[trait == "Alzheimer's disease" & hgnc_symbol %in% .herit.genes] %>%
    select(hgnc_symbol) %>%
    unique %>%
    left_join(.gene.sum) %>%
    na.omit %>%
    arrange(desc(nct)) %>%
    left_join(gene.info) %>%
    mutate(lb = pmax(transcript_start.hg19 - 1e6, 0)) %>%
    mutate(ub = transcript_end.hg19 + 1e6) %>%
    mutate(.bed = str_c(chr, ":", lb, "-", ub)) %>%
    mutate(.file = str_c("../result/step7/chr", chr, "_stat.bed.gz")) %>%
    mutate(query = str_c("tabix ", .file, " ", .bed, " -h")) %>%
    as.data.table

```


```{r}
plot.example <- function(g, .examples, .res = 1e4) {

    .hgnc <- unlist(.examples[g, "hgnc_symbol"])
    .chr <- unlist(.examples[g, "chr"])

    .dt <-
        fread(cmd = unlist(.examples[g, "query"]), header=TRUE) %>%
        filter(hgnc_symbol == .hgnc)

    .tss.loc <- floor(.examples[g, .(transcript_start.hg19)]/.res) %>% unlist

    .ct <- unique(.dt$celltype)[1]

    .dt.sum <- .dt[,
                   .(log10.pval = pmin(mean(-log10(p.val)), 10),
                     log10.pval.lb = pmin(min(-log10(p.val)), 10),
                     log10.pval.ub = pmin(max(-log10(p.val)), 10),
                     lodds = max(lodds)),
                   by = .(loc = floor(`stop`/.res), celltype)]
                          
    .dt.sum <- unique(.dt.sum[lodds > 0, .(celltype)]) %>%
        left_join(.dt.sum)

    .dt.sum[, celltype := factor(celltype, ct.order.dt$celltype)]

    if(nrow(.dt.sum) < 1) return(NULL)

    p1 <-
        .gg.plot(.dt.sum, aes(x = loc, y = sigmoid(lodds), size = lodds)) +
        ggtitle(.hgnc) +
        theme(title = element_text(size = 8)) +
        theme(strip.text = element_text(size = 6)) +
        theme(axis.text.x = element_blank()) +
        theme(axis.title.x = element_blank()) +
        ylab("Posterior\ninclusion\nprobablity") +
        facet_wrap(~celltype, nrow = 1) +
        scale_size_continuous(range=c(.5,1.5), guide="none") +
        geom_vline(aes(xintercept = loc), data = .dt.sum[lodds > 0],
                   colour = "red", size = .2) +
        geom_point(stroke = 0) +
        geom_vline(xintercept = .tss.loc, colour = "#009900", size = .1, lty = 2)
        geom_point(x = .tss.loc, y = 0, colour = "#009900")

    .x.scale <- scale_x_continuous("genomic location (kb) on chr" %&% .chr,
                                   labels = function(x) num.int(x*(.res/1e3)))


    p2 <-
        .gg.plot(.dt.sum, aes(x = loc, y = -log10.pval)) +
        theme(strip.text = element_blank()) +
        theme(axis.text.x = element_text(angle=90,vjust=1,hjust=1)) +
        scale_y_continuous("eQTL\np-value", labels = function(x) num.sci(10^x)) +
        .x.scale +
        facet_wrap(~celltype, nrow = 1) +
        geom_vline(xintercept = .tss.loc, colour = "#009900", size = .1, lty = 2) +
        geom_line(aes(y = -log10.pval.ub), size = .5) +
        geom_line(aes(y = -log10.pval.lb), size = .5) +
        geom_segment(aes(xend = loc, y = -log10.pval.lb, yend = -log10.pval.ub), size=.1) +
        geom_vline(aes(xintercept = loc), data = .dt.sum[lodds > 0],
                   colour = "red", size = .2)

    .mkdir(str_c(fig.dir, "/examples"))
    list(plt = p1/p2, file = fig.dir %&% "/examples/" %&% .hgnc %&% ".pdf")
}
```

```{r fig.width=10, fig.height=2}
for(g in 1:nrow(.examples)) {
    .out <- plot.example(g, .examples)
    if(!is.null(.out)) {
        print(.out$plt)
        .gg.save(.out$file, plot = .out$plt, width=10, height=2)
        cat("\n\n")
    }
}
```

